validator/src/journal/chain_commit_state.rs

```diff
@@ -21,6 +21,32 @@ use cpython;
 use cpython::ObjectProtocol;
 
 use batch::Batch;
+use block::Block;
+use journal::block_store::{BlockStore, IndexedBlockStore};
+
+#[derive(Debug, PartialEq)]
+pub enum ChainCommitStateError {
+    MissingDependency(String),
+    DuplicateTransaction(String),
+    DuplicateBatch(String),
+    Error(String),
+}
+
+pub struct ChainCommitState<BS: IndexedBlockStore> {
+    block_store: BS,
+    ancestor: Option<Block>,
+    uncommitted_batch_ids: Vec<String>,
+    uncommitted_txn_ids: Vec<String>,
+}
+
+fn check_no_duplicates(ids: &[&str]) -> Option<String> {
+    for (i, id1) in ids.iter().enumerate() {
+        if ids[0..i].contains(id1) || ids[i + 1..ids.len()].contains(id1) {
```

Comment:
seems like you can just check `i..ids.len()`.  If you are iterating from `0` on, you don't need to look at the slice before, you've already checked those as duplicates.

Reply:
This could be written as a double for loop, but is written with a for loop and contains.

`for (i, id) in ["1", "2", "1"]`

and then `if ["1"].contains("1") || ["1"].contains("1")` for the first loop through.